experiments/experiments.go

```diff
@@ -0,0 +1,195 @@
+	result := e.watcher.Get()
+	doc, ok := result.(Document)
+	if !ok {
+		return nil, fmt.Errorf("unexpected type %T", doc)
+	}
+	experiment, ok := doc[name]
+	if !ok {
+		return nil, ErrorUnknownExperiment(name)
+	}
+	if isSimpleExperiment(experiment.Type) {
+		return NewSimpleExperiment(experiment), nil
+	}
+	return nil, nil
+}
+
+type ParsedExperiment struct {
+	ExperimentVersion int       `json:"experiment_version"`
+	ShuffleVersion    string    `json:"shuffle_version"`
+	BucketVal         string    `json:"bucket_val"`
+	LogBucketing      bool      `json:"log_bucketing"`
+	Variants          []Variant `json:"variants"`
+	BucketSeed        string    `json:"bucket_seed"`
+}
+
+type ExperimentConfig struct {
+	// ID is the experiment identifier and should be unique for each experiment.
+	ID int `json:"id`
+	// Name is the experiment name and should be unique for each experiment.
+	Name           string           `json:"name"`
+	Owner          string           `json:"owner"`
+	Enabled        *bool            `json:"enabled"`
+	Version        string           `json:"version"`
+	Type           string           `json:"type"`
+	StartTimestamp int64            `json:"start_ts"`
+	StopTimestamp  int64            `json:"stop_ts"`
+	Experiment     ParsedExperiment `json:"experiment"`
+}
+
+type SimpleExperiment struct {
+	id           int
+	name         string
+	seed         string
+	hash         hash.Hash
+	numBuckets   int
+	enabled      bool
+	startTime    time.Time
+	endTime      time.Time
+	logBucketing bool
+	bucketVal    string
+	variantSet   VariantSet
+}
+
+// bucketVal == bucketing key, used to get the data
+func NewSimpleExperiment(experiment *ExperimentConfig) *SimpleExperiment {
+	shuffleVersion := experiment.Experiment.ShuffleVersion
+	if shuffleVersion == "" {
+		shuffleVersion = "None"
+	}
+
+	enabled := true
+	if experiment.Enabled != nil {
+		enabled = *experiment.Enabled
+	}
+
+	bucketSeed := experiment.Experiment.BucketSeed
+	if experiment.Experiment.BucketSeed == "" {
+		bucketSeed = fmt.Sprintf("%d.%s.%s", experiment.ID, experiment.Name, shuffleVersion)
+	}
+	return &SimpleExperiment{
+		id:         experiment.ID,
+		name:       experiment.Name,
+		seed:       bucketSeed,
+		hash:       sha1.New(),
+		bucketVal:  experiment.Experiment.BucketVal,
+		enabled:    enabled,
+		startTime:  time.Unix(experiment.StartTimestamp, 0),
+		endTime:    time.Unix(experiment.StopTimestamp, 0),
+		numBuckets: 1000,
+		variantSet: FromExperimentType(experiment.Type, experiment.Experiment.Variants, 0),
+	}
+}
+
+func (e *SimpleExperiment) Variant(bucketKey string) (string, error) {
+	if !e.isEnabled() {
+		return "", nil
+	}
+	bucket := e.calculateBucket(bucketKey)
+	return e.variantSet.ChooseVariant(bucket), nil
+}
+
+func (e *SimpleExperiment) calculateBucket(bucketKey string) int {
+	target := new(big.Int)
+	bucket := new(big.Int)
+	hashed := sha1.Sum([]byte(fmt.Sprintf("%s%s", e.seed, bucketKey)))
```

Comment:
I think you want to use `e.hash` instead of hardcoded `sha1.Sum` here?

Reply:
Do you mean allocating the hash in the constructor, i.e. `sha1.New()`?

Reply:
No you already did that before.

Reply:
on line 134

Reply:
Oh I see that needs to be removed because it seems to create a buffer that you can write to so it doesn't seem to be of purpose for reusing it, i.e.

```
h := sha1.New()
io.WriteString(h, "His money is twice tainted:")
io.WriteString(h, " 'taint yours and 'taint mine.")
fmt.Printf("% x", h.Sum(nil))
```

Reply:
You can call `Reset` if you need to reuse it, but that's not thread safe and cannot be used concurrently. If you need to reuse them concurrently, then the type should be `func() hash.Hash` (which `sha1.New` satisfies) instead.

Also if you don't think we'll ever use any hash other than sha1, then this is unnecessary and we can just hardcode sha1 instead.